[Messenger] Ability to distinguish retry and delay actions

theravel opened the pull request:
Added ability to distinguish retry and delay actions so that different "x-dead-letter-exchange" exchange name will be used in different scenarios.

| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | 

This is a bug which existed since v4.4. The following scenario is possible:

- There are two queues: `A` and `B`, both are bound to the same routing key via "topic" exchange (two different applications for example).
- A message is published to this routing key to "topic" exchange.
- Consumer of queue `A` handles it correctly and acknowledges the message.
- Consumer of queue `B` throws and exception and message goes to retry (for example to queue `delay_delays_key_5`).
- Once message expired in `delay_delays_key_5`, it is delivered again to both `A` and `B` (**again** consumed by consumer `A`).

Expected: behavior of consumer `B` should not cause message duplication to queue `A`.

It is required to make a change of name of temporary delay queue (otherwise "delay" and "retry" queues have incompatible declaration arguments). I left `queue_name_pattern` as is to keep settings of connection backward compatible, but changed internals of queue name construction.

nicolas-grekas commented:
> In my implementation I could not avoid minor incompatible change: queue_name_pattern has changed, so looks like this cannot be released under 4.4 or 5.0. Can we include it into 5.1 please?

Neither would this be allowed in master since it would break apps migrating to 5.2. Can we find another idea?

theravel commented:
> Neither would this be allowed in master since it would break apps migrating to 5.2. Can we find another idea?

@nicolas-grekas Thank you for the feedback. I changed it: now `queue_name_pattern` remains backward compatible, but I changed internals of queue name construction.

nicolas-grekas commented:
Yes please if this should be considered as a bugfix (I didn't check that part).

theravel commented:
@nicolas-grekas changed target to 4.4

Tobion commented on src/Symfony/Component/Messenger/Transport/AmqpExt/AmqpStamp.php:
```diff
@@ -45,7 +46,12 @@ public function getAttributes(): array
     }
 
-    public static function createFromAmqpEnvelope(\AMQPEnvelope $amqpEnvelope, self $previousStamp = null): self
+    public function isEnvelopeRedelivered(): bool
```
This is really confusing. AMQP also has the concept of redelivery. But this is not the same as the redelivery in the messenger component. Since this is the stamp to encode the amqp flags, this is really unexpected. So this needs a different name or not be exposed at all.

theravel replied:
This has been renamed

Tobion commented:
#36196 and #36212 already try to solve the same problem.

s-code commented:
@theravel i was able to solve this and few other problems without changing Messenger connection class, but it's little bit tricky. Here is an example https://github.com/s-code/amqp-extra-bundle. Take a look on ExtraAmqpFactory, ExtraAmqpSender, ExtraAmqpQueue, ExtraAmqpExchange.
As for me, slitting of retry and delay logic is a must have for amqp transport.

s-code commented on src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php:
```diff
@@ -240,15 +241,15 @@ private function publishOnExchange(\AMQPExchange $exchange, string $body, string
+        $queue = $this->createDelayQueue($delay, $routingKey, $isRetryAttempt);
         $queue->declareQueue(); // the delay queue always need to be declared because the name is dynamic and cannot be declared in advance
-        $queue->bind($this->connectionOptions['delay']['exchange_name'], $this->getRoutingKeyForDelay($delay, $routingKey));
+        $queue->bind($this->connectionOptions['delay']['exchange_name'], $this->getRoutingKeyForDelay($delay, $routingKey, $isRetryAttempt));
```
**retry** queue should not be bind to delay exchange

weaverryan commented:
On a high level, this makes sense to me:

If a message is in queue A, and I read it from queue A and it fails, then it should go back onto queue A only.

Messenger handles retries by actually ack'ing the original message and redelivering it. But ultimately, the desired behavior (I think) would be what I described above and what this PR does.

Though, apparently this and #38486 solve the same problem (is it the *exact* same problem?) but in different ways. @s-code could you comment on what your PR does versus this one?

Thanks :)

s-code commented:
@weaverryan you are right. This two pull requests solve the same problem. The difference is in next things:
1. Delay queue naming convention. 

-  #36864 requires its change from `delay_{exchange}_{routing_key}_{time}` to `delay_{exchange}_{routing_key}_{time}_(retry|delay)`. So as result we have `delay_messages_message1_500_retry`, `delay_messages_message1_500_delay` queues names for default configuration

- In  #38486 there is no changes in this convention. In case of retry `{exchange}` = default exchange = '', {routing_key} = origin queue name, because for default exchange `queue name` is using as `routing key`. So queues names look like `delay__queue1_500` and `delay_messages_message1_500`

2. #38486 allows to use default exchange for delay and retry logic. If in configuration we define delay.exchange_name as '', no extra ("delays") exchange will be created.  For me it's more elegant and clean solution

theravel committed:
[Messenger] Ability to distinguish retry and delay actions

theravel commented:
Dear Symfony team,

I have:
* Fixed conflicts and rebased to fresh `5.x`
* Fixed confusion about `redelivery` term

Can you take a look please?

theravel commented:
@Tobion could you have a look please?

nicolas-grekas commented on src/Symfony/Component/Messenger/Bridge/Amqp/Transport/AmqpStamp.php:
```diff
@@ -45,7 +46,12 @@ public function getAttributes(): array
     }
 
-    public static function createFromAmqpEnvelope(\AMQPEnvelope $amqpEnvelope, self $previousStamp = null): self
+    public function isEnvelopeRetried(): bool
```
would it be more clear to name this isRetryAttempt as you did in Connection? or just isRetry (in Connection also then)?

theravel replied:
I agree, and thank you for review!
I renamed it everywhere to `isRetryAttempt`, feel free to take a look once again please.

nicolas-grekas commented:
(please add a changelog entry in the bridge)

theravel commented:
@nicolas-grekas Changelog entry added

nicolas-grekas commented on src/Symfony/Component/Messenger/Bridge/Amqp/CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@ CHANGELOG
 
  * Deprecated the `prefetch_count` parameter, it has no effect and will be removed in Symfony 6.0.
  * `AmqpReceiver` implements `QueueReceiverInterface` to fetch messages from a specific set of queues.
+ * Fixed message retry, which caused initial message redelivery to all consumers in setup with multiple active consumers.
```
` * Add ability to distinguish retry and delay actions`

theravel replied:
Updated to your phrasing, thank you

nicolas-grekas commented on src/Symfony/Component/Messenger/Bridge/Amqp/CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@ CHANGELOG
 
  * Deprecated the `prefetch_count` parameter, it has no effect and will be removed in Symfony 6.0.
  * `AmqpReceiver` implements `QueueReceiverInterface` to fetch messages from a specific set of queues.
+ * Add ability to distinguish retry and delay actions.
```
no final dot please (to match our recent policy related to changelogs)

theravel replied:
Removed final dot. (I was looking at the previous records)

Nyholm commented on src/Symfony/Component/Messenger/Bridge/Amqp/Transport/AmqpStamp.php:
```diff
@@ -45,7 +46,12 @@ public function getAttributes(): array
     }
 
-    public static function createFromAmqpEnvelope(\AMQPEnvelope $amqpEnvelope, self $previousStamp = null): self
+    public function isRetryAttempt(): bool
```
Lets add this new method below the named factory method. (createFromAmqpEnvelope)

theravel replied:
Done

Nyholm commented on src/Symfony/Component/Messenger/Bridge/Amqp/Transport/AmqpStamp.php:
```diff
@@ -45,7 +46,7 @@ public function getAttributes(): array
     }
 
-    public static function createFromAmqpEnvelope(\AMQPEnvelope $amqpEnvelope, self $previousStamp = null): self
+    public static function createFromAmqpEnvelope(\AMQPEnvelope $amqpEnvelope, self $previousStamp = null, string $retryRoutingKey = null): self
```
We are using `$amqpReceivedStamp->getQueueName()` to populate the `$retryRoutingKey`, is that really correct? 

A routing key is not the same as the queue name. 

(It is probably correct, but I just need you to clarify.)

theravel replied:
This is correct. We want to distinguish:
* delay scenario, where message should be sent to original exchange to original routing key (so that all consumers can get it)
* retry scenario, where message should be sent to default (direct) exchange to routing key which matches queue name (in this case it will be received in one queue only)

Nyholm replied:
Are we sure the routing key always matches the queue name?

theravel replied:
For direct exchange in AMQP 0-9-1 yes:
> The default exchange is a direct exchange with no name (empty string) pre-declared by the broker. It has one special property that makes it very useful for simple applications: every queue that is created is automatically bound to it with a routing key which is the same as the queue name.

See https://www.rabbitmq.com/tutorials/amqp-concepts.html

theravel replied:
@Nyholm can you review please?

Nyholm reviewed (commented):
Im happy with this. Just one question

MichaelKubovic commented:
I believe it is always the case with direct exchange.

> The default exchange is a direct exchange with no name (empty string) pre-declared by the broker. It has one special property that makes it very useful for simple applications: every queue that is created is automatically bound to it with a routing key which is the same as the queue name.

https://www.rabbitmq.com/tutorials/amqp-concepts.html


> On 21 Apr 2021, at 15:40, Tobias Nyholm ***@***.***> wrote:
> 
> @Nyholm commented on this pull request.
> 
> In src/Symfony/Component/Messenger/Bridge/Amqp/Transport/AmqpStamp.php:
> 
> > @@ -45,7 +46,7 @@ public function getAttributes(): array
>          return $this->attributes;
>      }
>  
> -    public static function createFromAmqpEnvelope(\AMQPEnvelope $amqpEnvelope, self $previousStamp = null): self
> +    public static function createFromAmqpEnvelope(\AMQPEnvelope $amqpEnvelope, self $previousStamp = null, string $retryRoutingKey = null): self
> Are we sure the routing key always matches the queue name?
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

Nyholm commented:
Thank you @theravel and @MichaelKubovic for confirming.

Nyholm commented:
Thank your for this feature. Please update the docs too =)
